src/tests/performance/blocks.rs

```diff
@@ -0,0 +1,222 @@
+    const REQUESTS: usize = 100;
+    const REQUEST_TIMEOUT: Duration = Duration::from_millis(100);
+    // number of concurrent peers to test (zcashd hardcaps `max_peers` to 873 on my machine)
+    let peer_counts = vec![
+        1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 200, 300, 500, 750, 800,
+    ];
+
+    let mut table = Table::new();
+    table
+        .load_preset(UTF8_FULL)
+        .apply_modifier(UTF8_ROUND_CORNERS)
+        .set_header(vec![
+            Cell::new("Peers").set_alignment(CellAlignment::Center),
+            Cell::new("Requests").set_alignment(CellAlignment::Center),
+            Cell::new("Min (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("Max (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("stddev (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("10% (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("50% (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("75% (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("90% (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("99% (ms)").set_alignment(CellAlignment::Center),
+            Cell::new("Time (s)").set_alignment(CellAlignment::Center),
+            Cell::new("Request/s").set_alignment(CellAlignment::Center),
+        ]);
+
+    // Start node seeded with initial testnet blocks,
+    // with max peers set so that our peers should never be rejected.
+    let mut node: Node = Default::default();
+    node.initial_action(Action::SeedWithTestnetBlocks {
+        socket_addr: new_local_addr(),
+        block_count: 3,
+    })
+    .max_peers(peer_counts.iter().max().unwrap() * 2 + 10)
+    .start()
+    .await;
+    let node_addr = node.addr();
+
+    for peers in peer_counts {
+        // create N peer nodes which send M requests's as fast as possible
+        let mut peer_handles = Vec::with_capacity(peers);
+
+        let test_start = tokio::time::Instant::now();
+
+        for _ in 0..peers {
+            // We want different blocks for consecutive requests, in order to determine if the node
+            // has skipped a request or to tell if the reply is in response to a timed out request.
+            //
+            // We also store the Block, in order to compare to the reply.
+            let mut requests = Block::initial_testnet_blocks()
+                .into_iter()
+                .map(|block| {
+                    (
+                        Message::GetData(Inv::new(vec![block.inv_hash()])),
+                        Box::new(block),
+                    )
+                })
+                .collect::<VecDeque<_>>();
+
+            peer_handles.push(tokio::spawn(async move {
+                let mut stream = initiate_handshake(node_addr).await.unwrap();
+
+                let filter = MessageFilter::with_all_auto_reply();
+
+                let mut latencies = Vec::with_capacity(REQUESTS);
+                for _ in 0..REQUESTS {
```

Comment:
This implies the same requests are made in cycles of 3? I wonder if `zcashd` or `zebra` would want throttle the same request from the same address (as this test runs fine, I'd assume they don't currently).

Reply:
Yes that's currently the case. Would be trivial to extend that to be random, but with only three blocks the combinations are somewhat limited.